README.md
MAST
====

Matlab Toolbox for SPTK


==DESCRIPTION==

MAST is a small MATLAB toolbox that attempts to alleviate the pain of using SPTK in conjunction with MATLAB. It consists of a set of objects to interface with the local SPTK installation and tries to be as intuitive as possible using object-oriented paradigms.

It's in very early stages of development, with the philosophy of building a small set of functionality with as much forward-thinking as possible, and iteratively increase the functionality and scope based on user feedback and experience.


==USAGE INSTRUCTIONS==

[Windows is not supported yet, sorry! Let me know if you need it!]

At the heart of the toolbox is the sptk class. It serves both as a static file to set the environment variables, and (as an object) the interface to the transparent system calls made to the local SPTK binaries.
To use the toolbox, you should first open sptk.m and make sure the path properties are set correctly for your system. The defaults should mostly serve on a regular UNIX system, but if you want to use pitch extraction using the HTS method, make sure you point getf0tcl to your copy of getF0.tcl. (Note that the RAPT option in SPTK's pitch binary is the same algorithm used in getF0.tcl)
Analysis parameters are set to be consistent for analysis at 48kHz. For convenient access to parameters for 16kHz, look at the end of this file.

The data classes in the toolbox, by inheritance order, with their methods (some of which are overloaded) are:
speechData : read(), write(), isPrecision()
  |=> timeDomainData : plot(), length(), extract(), min(), max(), minmax()
		|=> pitchData : plot()
	|=> freqDomainData : read(), plot(), min(), max(), minmax()
		|=> mgcData : plot(), read()
		
Processing of the data is done through methods of the sptk class. Currently available processing methods are:
sptk
	|=> extractMgc()
	|=> filterMlsadf()
	|=> extractPitchGetF0()
  |=> extractPitchRAPT()
  |=> extractPitchSWIPE()
	|=> exciteFromPitch()
	
	
Some examples that might be encountered in a typical workflow are demonstrated instead of an extensive manual:
>> s = sptk;
>> speech = timeDomainData(s,rawFile,'int16');
>> speech.plot %alternatively plot(speech)
>> mgc = s.extractMgc(speech);
>> residual = s.filterMlsadf(speech,mgc,'inverse');
>> residual.write(residualPath);
>> pitch = s.extractPitchHts(speech);
>> pulse = s.exciteFromPitch(pitch);
>> synth = s.filterMlsadf(pulse,mgc);
>> synth.plot
>> synth.write(synthPath,'int16');


There are some more features, and the user is invited to explore the code and get further information from comments.


==TO-DO LIST==

-Better documentation
-Add .wav reading method to timeDomainData (same syntax, detect from extension and/or MIME type)
-Add getters and setters to sptk to ensure the analysis parameters are meaningful and consistent with each other
-Integrate mglsadf calls into sptk.filterMlsadf()
-Get user feedback, polish and expand the feature set :)


==ACKNOWLEDGMENTS==
Big thanks to Alexis Moinet for his design suggestions. They immediately snapped things into focus in my head and saved me days of wallowing in messy code. 


==================================================================
MAST is written by Onur Babacan at the TCTS Lab in University of Mons, Belgium, in 2012-2013.











=====Replace the block in sptk.m with this one for processing at 16kHz 

        %SPTK Analysis parameters for 16kHz
        fs = 16000;
        frameLength = 400;
        frameShift = 80;
        windowType = 1; % 0: Blackman 1: Hamming 2: Hanning
        normalize = 1; % 0: none 1: by power 2: by magnitude
        fftLength = 512;
        mgcAlpha = 0.42;
        mgcGamma = 0;
        mgcOrder = 24;
        mgcLsp = 0; %convert MGC coefficients into MGC-LSP form
        lnGain = 1; %use logarithmic gain instead of linear gain
        noiseMask = 50; %standard deviation of white noise to mask noises in f0 extraction


changelog.txt
MAST Alpha 0.04

-Fixed a minor bug in timeDomainData plotting (error when tdd.sourceFilename was empty)
-For unavoidable failures of "determinant too small" in mgcep (i.e. adjusting -e and -f doesn't work), 
added a feature where MAST tries MGC extraction at a higher order, then, if successful converts those 
coefficients to the original target order.
-Added properties to sptk class for using the "-e" and "-f" options in mgcep (through sptk.extractMgc), 
sptk.mgcPeriodogramOffset and sptk.mgcMinimumDeterminant, respectively.



MAST Alpha 0.03
-Renamed pitch extraction functions to be correctly named according to algorithm
	|=> renamed sptk.extractPitchHTS() to sptk.extractPitchGetF0()
	|=> added sptk.extractPitchRAPT()
	*These functions both use the RAPT algorithm, but different implementations.
	sptk.extractPitchGetF0 requires the getf0.tcl script, which can be found e.g. in the HTS training scripts
-Added the option to use a specified voiced/unvoiced threshold in sptk.extractPitchSWIPE()
The value is stored in sptk.thresholdSWIPE with the default of 0.3
-Added "Speak Aloud" feature, which displays the calls made to the system in the MATLAB command window. Set sptk.speakAloud to 1 to enable.



MAST Alpha 0.02
-Fixed isPrecision not being found in freqDomainData and mgcData (Thanks Alexis!)
-Fixed the error in plot titles due to improper referencing of the cell array. Titles now display the last element of the sourceFilename array recursively. (Thanks Alexis!)
-Turned off the LaTeX interpreter in plot titles to fix display of filenames with underscores.



MAST Alpha 0.01
-First release!


mgcData.m
%Usage: 
%m = mgcData(s)
%m = mgcData(fs,frameLength,frameShift,numberOfBins,order,alpha,gamma,windowTypeString)
%
%This usage only works with an instance of mgcData (i.e. it's not static)
%m = m.read(filePath,precision)

classdef mgcData < freqDomainData
    
    properties
        order
        alpha
        gamma
    end
    
    methods
        function obj = mgcData(varargin)
            obj = obj@freqDomainData();
            if(isempty(varargin))
                
            elseif(length(varargin) == 1 && isa(varargin{1},'sptk'))
                s = varargin{1};
                obj = mgcData(s.fs,s.frameLength,s.frameShift,s.fftLength,s.mgcOrder,s.mgcAlpha,s.mgcGamma,s.windowTypeChar);
            elseif(length(varargin) == 8)
                if(length(varargin{1})==1)
                    obj.fs = varargin{1};
                else
                    error('Sampling frequency must be a single value');
                end
                if(length(varargin{2})==1)
                    obj.frameLength = varargin{2};
                else
                    error('Frame length must be a single value');
                end
                if(length(varargin{3})==1)
                    obj.frameShift = varargin{3};
                else
                    error('Frame shift must be a single value');
                end
                if(length(varargin{4})==1)
                    obj.nBins = varargin{4};
                else
                    error('Number of bins must be a single value');
                end
                if(length(varargin{5})==1)
                    obj.order = varargin{5};
                else
                    error('Order must be a single value');
                end
                if(length(varargin{6})==1)
                    obj.alpha = varargin{6};
                else
                    error('Alpha must be a single value');
                end
                if(length(varargin{7})==1)
                    obj.gamma = varargin{7};
                else
                    error('Gamma must be a single value');
                end
                if(isa(varargin{8},'char'))
                    obj.windowType = varargin{8};
                else
                    error('Window type must be a char, and one of the known window handles (see help window)');
                end
                obj.data = zeros(obj.order+1,1);
                obj.maxColorValue = []; 
            else
                error('Unrecognized input argument pattern')
            end
        end

        
        function plot(obj,varargin)
            if(length(obj) > 0) %#ok<ISMT>
                if(isempty(varargin))
                    t = (1:obj.length)*(obj.frameShift/obj.fs);
                elseif(length(varargin)==2)
                    boundaries = varargin{2};
                    if(boundaries(1)>0 && boundaries(2)<obj.length)
                        t = (boundaries(1):boundaries(2))*(1/obj.fs)*obj.frameShift;
                    end
                end
                
                image(obj.data)
                
                set(gca,'YDir','normal');
                
                set(gca,'XTick',1:100:obj.length);
                set(gca,'XTickLabel',t(1:100:obj.length));
                
                y = 0:obj.order;
                set(gca,'YTick', 0:obj.order);
                set(gca,'YTickLabel',y);
                
                if(isempty(obj.maxColorValue))
                    if(max(max(obj.data)))
                        colormap(jet( ceil(max(max(obj.data))) ));
                    else
                        colormap(jet(512));
                    end
                else
                    colormap(jet(obj.maxColorValue))
                end
                
                xlabel('Time (s)');
                ylabel('Mel-cepstral coefficients');
            end
        end
        
        
        function obj = read(obj,filePath,precision)
            
            if( ~obj.isPrecision(precision) )
                error([precision ' is not recognized as a precision (data type)']);
            end
            fp = fopen(filePath,'r');
            if(fp < 0)
                error([filePath ' could not be opened']);
            end
            obj.data = fread(fp, precision);
            obj.data = reshape( obj.data,obj.order+1,length(obj.data)/(obj.order+1) );
            fclose(fp);
            obj.sourceFilename = cell(0);
            obj.sourceFilename{end+1} = filePath;
        end
        
        %No need to overload write() because fwrite() called from speechData.write
        %implements the exact complement of mgcData.read();
        
    end
    
    
    
    
end

pitchData.m
%A simple class for storing pitch trajectories

classdef pitchData < timeDomainData
    
    properties
        frameShift
        frameLength
    end
    
    methods
        
        function obj = pitchData(varargin)
            obj = obj@timeDomainData();
            if(length(varargin) ~= 1 || ~isa(varargin{1},'sptk'))
                error('Class pitchData must be instantiated only with one sptk object')
            end
            obj.fs = varargin{1}.fs;
            obj.data = [];
            obj.sourceFilename = [];
            obj.frameShift = varargin{1}.frameShift;
            obj.frameLength = varargin{1}.frameLength;
        end
        
                    %Overload plot()
            function [varargin] = plot(obj,varargin)
                if(obj.length > 0)
                    t = (1:obj.length)*obj.frameShift/obj.fs;
    
                    if isempty(varargin)
                        p = plot(t,obj.data);
                        xlabel('Time (s)')
                        ylabel('Pitch (Hz)')
                        
                        %Recursively find the last filename in the sourceFilename "tree"
                        c = obj.sourceFilename{end};
                        while(iscell(c))
                            c = c{end};
                        end
                        h = title(c);
                        set(h,'interpreter','none');
                    else
                        if(length(varargin) == 1)
                            boundaries = varargin{1};
                            plot( t(boundaries(1):boundaries(2)) , obj.data(boundaries(1):boundaries(2)) );
                        else
                            error('One start point and one end point must be given');
                        end
                        
                    end
                    varargin{1} = p;
                else
                    error('There is no data in the object');
                end
            end
    end
    
    
    
    
    

    
    
    
end


speechData.m
%Abstract class from which other data classes inherit

classdef speechData
    
    properties(Abstract = true)
        fs
        data
        sourceFilename
    end
    
    methods
        function obj = read(obj,filePath,precision)
            %Check if precision is recognized
            if( ~obj.isPrecision(precision) )
                error([precision ' is not recognized as a precision (data type)']);
            end
            
            fp = fopen(filePath,'r');
            if(fp < 0)
                error([filePath ' could not be opened']);
            end
            
            obj.data = fread(fp, precision);
            fclose(fp);
            obj.sourceFilename = cell(0);
            obj.sourceFilename{end+1} = filePath;
        end
        
        function write(obj,filePath,precision)
            %Check if precision is recognized
            if( ~obj.isPrecision(precision) )
                error([precision ' is not recognized as a precision (data type)']);
            end
            fp = fopen(filePath,'w');
            if(fp < 0)
                error([filePath ' could not be opened']);
            end
            fwrite(fp, obj.data, precision);
            fclose(fp);
        end
        
        function b = isPrecision(obj,precisionString)
            if( strcmpi(precisionString, 'int8') || strcmpi(precisionString, 'uint8') || strcmpi(precisionString, 'int16') || strcmpi(precisionString, 'uint16') || strcmpi(precisionString, 'int32') || strcmpi(precisionString, 'uint32') || strcmpi(precisionString, 'int64') || strcmpi(precisionString, 'uint64') || strcmpi(precisionString, 'float') || strcmpi(precisionString, 'single') || strcmpi(precisionString, 'double') )
                b = 1;
            else
                b = 0;
            end
        end
        
    end
end

sptk.m
classdef sptk
    properties
        
        sptkDir = '/usr/local/SPTK/bin';
        
        bc = '/usr/bin/bc';
        wc = '/usr/bin/wc';
        tclsh = '/usr/local/bin/tclsh';
        getf0tcl = '/Users/Work/Sources/MATLAB/Toolboxes/getf0.tcl';
        
%         %SPTK Analysis parameters/
%         fs = 48000;
%         frameLength = 1200;
%         frameShift = 240;
%         windowType = 1; % 0: Blackman 1: Hamming 2: Hanning
%         normalize = 1; % 0: none 1: by power 2: by magnitude
%         fftLength = 2048;
%         mgcAlpha = 0.55;
%         mgcGamma = 0;
%         mgcOrder = 34;
%         mgcLsp = 0; %convert MGC coefficients into MGC-LSP form
%         lnGain = 1; %use logarithmic gain instead of linear gain
%         noiseMask = 50; %standard deviation of white noise to mask noises in f0 extraction


        %SPTK Analysis parameters for 16kHz
        fs = 16000;
        frameLength = 400;
        frameShift = 80;
        windowType = 1; % 0: Blackman 1: Hamming 2: Hanning
        normalize = 1; % 0: none 1: by power 2: by magnitude
        fftLength = 512;
        mgcAlpha = 0.42;
        mgcGamma = 0;
        mgcOrder = 24;
        mgcLsp = 0; %convert MGC coefficients into MGC-LSP form
		mgcPeriodogramOffset = 0;
		mgcMinimumDeterminant = 1e-6;
        lnGain = 1; %use logarithmic gain instead of linear gain
        noiseMask = 50; %standard deviation of white noise to mask noises in f0 extraction

        
        mlsadfFilterWithoutGain = 0; %-k option of mgcep
        
        f0Min = 110;
        f0Max = 280;
        thresholdSWIPE = 0.3;
        shift = 3;
        
        precision = 'single'; %used for file i/o
        
        verbose = 0;
        speakAloud = 0; %Set to 1 to display terminal commands in the MATLAB command window
        
    end
    
    properties (Dependent = true);
        
        swab
        x2x
        frame
        window
        mgcep
        lpc2lsp
        step
        merge
        vstat
        nrand
        sopr
        excite
        vopr
        nan
        mlsadf
        pitch
		mgc2mgc
        
        tmpDir
        windowTypeChar
        
    end
    
    methods
        %==Dependent property constructors==
        function swab = get.swab(s)
            swab = [s.sptkDir , '/swab'];
        end
        function x2x = get.x2x(s)
            x2x = [s.sptkDir , '/x2x'];
        end
        function frame = get.frame(s)
            frame = [s.sptkDir , '/frame'];
        end
        function window = get.window(s)
            window = [s.sptkDir , '/window'];
        end
        function mgcep = get.mgcep(s)
            mgcep = [s.sptkDir , '/mgcep'];
        end
        function lpc2lsp = get.lpc2lsp(s)
            lpc2lsp = [s.sptkDir , '/lpc2lsp'];
        end
        function step = get.step(s)
            step = [s.sptkDir , '/step'];
        end
        function merge = get.merge(s)
            merge = [s.sptkDir , '/merge'];
        end
        function vstat = get.vstat(s)
            vstat = [s.sptkDir , '/vstat'];
        end
        function nrand = get.nrand(s)
            nrand = [s.sptkDir , '/nrand'];
        end
        function sopr = get.sopr(s)
            sopr = [s.sptkDir , '/sopr'];
        end
        function excite = get.excite(s)
            excite = [s.sptkDir , '/excite'];
        end
        function vopr = get.vopr(s)
            vopr = [s.sptkDir , '/vopr'];
        end
        function nan = get.nan(s)
            nan = [s.sptkDir , '/nan'];
        end
        function mlsadf = get.mlsadf(s)
            mlsadf = [s.sptkDir , '/mlsadf'];
        end
        function pitch = get.pitch(s)
            pitch = [s.sptkDir , '/pitch'];
		end
		function mgc2mgc = get.mgc2mgc(s)
            mgc2mgc = [s.sptkDir , '/mgc2mgc'];
        end
        
        function tmpDir = get.tmpDir(s) %#ok<MANU>
            if(isunix())
                tmpDir = '/tmp';
            elseif(ispc())
                tmpDir = 'C:\TEMP';
            end
        end
        function wtchar = get.windowTypeChar(s)
            switch s.windowType
                case 0
                    wtchar = 'blackman';
                case 1
                    wtchar = 'hamming';
                case 2
                    wtchar = 'hanning';
            end
        end
        %==Dependent property constructors==
        
        %SPTK Methods
        function mgc = extractMgc(s,tdd)
            %Usage
            %mgc = s.extractMgc(tdd)ex
            
            if(~isa(tdd,'timeDomainData'))
                error('Input must implement class timeDomainData')
            end
            
            %Paths to temporary files
            tmpInput = [s.tmpDir  filesep 'MAST_extractMgc_tmpdata'];
            tmpOutput = [s.tmpDir  filesep 'MAST_extractMgc_tmpdataout'];
            if(s.verbose)
                display('MAST:Extracting MGCs using mgcep...')
            end
            tdd.write(tmpInput,'float');
			
			[status, result] = systemCallMAST(s, [s.frame ' -l ' num2str(s.frameLength) ' -p ' num2str(s.frameShift) ' ' tmpInput ' | ' s.window ' -l ' num2str(s.frameLength) ' -L ' num2str(s.fftLength) ' -w ' num2str(s.windowType) ' -n 1 | ' s.mgcep ' -f ' num2str(s.mgcMinimumDeterminant,'%.16f') ' -e ' num2str(s.mgcPeriodogramOffset,'%.16f') ' -a ' num2str(s.mgcAlpha) ' -m ' num2str(s.mgcOrder) ' -l ' num2str(s.fftLength) ' > ' tmpOutput] );
			if(status)
				display('MAST: Retrying MGC extraction with higher order because: ')
				disp([result])
				display(['MAST: Results will be converted back to the original order'])
				mgcExtractionSuccessful = 0;
				
				newTmpInput = [s.tmpDir  filesep 'MAST_extractMgc_retry_tmpdata'];
				
% 				[statusMove resultMove] = movefile(tmpOutput , newTmpInput);
% 				if(statusMove)
% 					error(resultMove)
% 				end
				
				numberOfTries = 1;
				maxRetries = 5;
				while(numberOfTries <= maxRetries && ~mgcExtractionSuccessful)
					newMgcOrder = s.mgcOrder + numberOfTries;
					[statusRetry, resultRetry] = systemCallMAST(s, [s.frame ' -l ' num2str(s.frameLength) ' -p ' num2str(s.frameShift) ' ' tmpInput ' | ' s.window ' -l ' num2str(s.frameLength) ' -L ' num2str(s.fftLength) ' -w ' num2str(s.windowType) ' -n 1 | ' s.mgcep ' -f ' num2str(s.mgcMinimumDeterminant,'%.16f') ' -e ' num2str(s.mgcPeriodogramOffset,'%.16f') ' -a ' num2str(s.mgcAlpha) ' -m ' num2str(newMgcOrder) ' -l ' num2str(s.fftLength) ' > ' newTmpInput] );
					if(statusRetry)
						numberOfTries = numberOfTries + 1;
					else
						mgcExtractionSuccessful = 1;
						[statusConvert, resultConvert] = systemCallMAST(s, [s.mgc2mgc ' -m ' num2str(newMgcOrder) ' -a ' num2str(s.mgcAlpha) ' -g ' num2str(s.mgcGamma) ' -M ' num2str(s.mgcOrder) ' -A ' num2str(s.mgcAlpha) ' -G ' num2str(s.mgcGamma) ' < ' newTmpInput ' > ' tmpOutput ]);
						if(statusConvert)
							error(resultConvert)
						end
					end
				end
			end
          
            mgc = mgcData(s);
            mgc = mgc.read(tmpOutput,'float');
            %While processing files with this toolbox, proper usage is:
            %mgc.sourceFilename{end+1} = tdd.sourceFilename;
            %But we know this is the origin of this object, so it converges to the next line in this case
            mgc.sourceFilename{1} = tdd.sourceFilename;
            
            %Clean up temporary files
            if(isunix())
                systemCallMAST(s,['rm -f ' tmpInput]);
                systemCallMAST(s,['rm -f ' tmpOutput]);
            elseif(ispc())
                %TODO: For Windows
            end
            if(s.verbose)
                display('MAST:MGC extraction using mgcep complete.')
            end
        end
        
        function tddFiltered = filterMlsadf(s,tdd,mgc,varargin)
            %Usage
            %tddFiltered = s.filterMlsadf(tdd,mgc)
            %tddFiltered = s.filterMlsadf(tdd,mgc,'inverse') or tddFiltered = s.filterMlsadf(tdd,mgc,'i')
            
            
            %Resolve whether filtering without gain is requested
            if(s.mlsadfFilterWithoutGain)
                k = ' -k ';
            else
                k = [];
            end
            
            %Resolve whether inverse filtering is requested
            v = [];
            if(length(varargin) == 0) %#ok<*ISMT>
            elseif(length(varargin) == 1)
                if(isa(varargin{1},'char'))
                    if(strcmpi(varargin{1},'inverse') || strcmpi(varargin{1},'i'))
                        v = ' -v ';
                    else
                        error(['Unrecognized input argument' varargin{1}])
                    end
                else
                    error('Optional input argument must be of type char');
                end
            else
                error('Too many input arguments')
            end
            
            tmpMgc = [s.tmpDir  filesep 'MAST_mlsadf_tmpmgc'];
            tmpExc = [s.tmpDir  filesep 'MAST_mlsadf_tmpexc'];
            tmpSynth = [s.tmpDir filesep 'MAST_mlsadf_tmpsynth'];
            
            tdd.write(tmpExc,'float');
            mgc.write(tmpMgc,'float');
            
            if(s.verbose)
                if(isempty(v))
                    display('MAST:Filtering with mlsadf...')
                else
                    display('MAST:Inverse filtering with mlsadf...')
                end
            end
            
            [status, result] = systemCallMAST(s, [s.mlsadf ' -p ' num2str(s.frameShift) ' -a ' num2str(s.mgcAlpha) ' -m ' num2str(s.mgcOrder) k v ' ' tmpMgc ' ' tmpExc ' > ' tmpSynth] );
            if(status)
                error(result);
            end
            
            tddFiltered = timeDomainData(s,tmpSynth,'float');
            %because sourceFilename{end} will be the temporary file path in tmpSynth,
            %we can directly overwrite it instead of removing it and concatenating our
            %real source file paths
            tddFiltered.sourceFilename{end} = tdd.sourceFilename;
            tddFiltered.sourceFilename{end+1} = mgc.sourceFilename;
            
            %Clean up files
            if(isunix())
                systemCallMAST(s,['rm -f ' tmpMgc]);
                systemCallMAST(s,['rm -f ' tmpExc]);
                systemCallMAST(s,['rm -f ' tmpSynth]);
            elseif(ispc())
                %TODO: For Windows
            end
            
            if(s.verbose)
                if(isempty(v))
                    display('MAST:Filtering with mlsadf complete.')
                else
                    display('MAST:Inverse filtering with mlsadf complete.')
                end
            end
            
        end
        
        function pitch = extractPitchGetF0(s,tdd)
            %speechSignal values are floats in [0,1]

            
            if(~isa(tdd,'timeDomainData'))
                error('Input object is not of type timeDomainData')
            end
            tmp = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_tmp'];
            tmpRaw = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_tmp.raw'];
            tmpHead = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_tmp.head'];
            tmpTail = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_tmp.tail'];
            lf0File = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_lf0'];
            rawFile = [s.tmpDir filesep 'sptkExtractPitchRAPTgetf0_tmprawfile.raw'];
            
            speechSignal = tdd.data;
            if(size(speechSignal,2)==2)
                warning('Using column 1 of 2-column data') %#ok<WNTAG>
                speechSignal = speechSignal(:,1);
            end
            speechSignal = speechSignal ./ max(abs(speechSignal));
            
            speechSignal = speechSignal .* (precisionScalingFactor('int16')/precisionScalingFactor(class(speechSignal)));
            
            fp = fopen(rawFile,'w');
            fwrite(fp,speechSignal,'int16');
            fclose(fp);
            
            if(s.verbose)
                display('MAST:Pitch extraction using the RAPT algorithm (getf0.tcl)...')
            end
            
            [status, result] = systemCallMAST(s, [ s.step ' -l ' num2str(s.frameShift) ' -v 0.0 | ' s.x2x ' +fs > ' tmpHead ] );
            if(status)
                error(result);
            end
            [status, result] = systemCallMAST(s, [ s.step ' -l ' num2str(s.frameLength) ' -v 0.0 | ' s.x2x ' +fs > ' tmpTail ] );
            if(status)
                error(result);
            end
            
            if(isunix())
                [status, result] = systemCallMAST(s, [ 'cat ' tmpHead ' ' rawFile ' ' tmpTail ' | ' s.x2x ' +sf > ' tmp] );
                if(status)
                    error(result);
                end
                [status, result] = systemCallMAST(s, [  s.x2x ' +fa ' tmp ' | ' s.wc ' -l' ] );
                lineCount = num2str(str2num(result)); %#ok<ST2NM>
                
                if(status)
                    error(result);
                end
            else
                %TODO:For Windows
            end
            
[status, result] = systemCallMAST(s, [ s.nrand ' -l ' lineCount ' | ' s.sopr ' -m ' num2str(s.noiseMask) ' | ' s.vopr ' -a ' tmp ' > ' tmpRaw ] );
            if(status)
                error(result);
            end
            
            fp = fopen(tmpRaw,'r');
            floatData = fread(fp,'float');
            fclose(fp);
            floatData = floatData./(max(abs(floatData)))*32768;
            fp = fopen(tmpRaw,'w');
            fwrite(fp,floatData,'int16');
            fclose(fp);
            
            [status, result] = systemCallMAST(s, [ s.tclsh ' ' s.getf0tcl ' -l -lf0 -H ' num2str(s.f0Max) ' -L ' num2str(s.f0Min) ' -p ' num2str(s.frameShift) ' -r ' num2str(s.fs) ' ' tmpRaw ' | ' s.x2x ' +af > ' lf0File ] );
            if(status)
                error(result);
            end
            
            %Prepare object to be returned
            pitch = pitchData(s);
            pitch = pitch.read(lf0File,'float');
            %Convert from log(Hz) to Hz
            pitch.data = exp(pitch.data);
            pitch.sourceFilename = tdd.sourceFilename;
            pitch.fs = s.fs;
            
            
            %Clean up
            if(isunix())
                [status, result] = systemCallMAST(s,[ 'rm -f ' tmp ' ' tmpRaw ' ' tmpHead ' ' tmpTail ' ' lf0File ' ' rawFile ]); %remove temporary files
                %TODO: For Windows
            end
            if(status)
                error(result)
            end
            
            if(s.verbose)
                display('MAST:Pitch extraction using the RAPT algorithm (getf0.tcl) complete.')
            end
            
        end
        
         function pitch = extractPitchRAPT(s,tdd)
            %speechSignal values are floats in [0,1]

            
            if(~isa(tdd,'timeDomainData'))
                error('Input object is not of type timeDomainData')
            end

            f0File = [s.tmpDir filesep 'sptkExtractPitchRAPT_f0'];
            rawFile = [s.tmpDir filesep 'sptkExtractPitchRAPT_tmprawfile.raw'];
            
            speechSignal = tdd.data;
            if(size(speechSignal,2)==2)
                warning('Using column 1 of 2-column data') %#ok<WNTAG>
                speechSignal = speechSignal(:,1);
            end
            speechSignal = speechSignal ./ max(abs(speechSignal));
			speechSignal = speechSignal .* (precisionScalingFactor('int16')/precisionScalingFactor(class(speechSignal)));

			
			
            fp = fopen(rawFile,'w');
            fwrite(fp,speechSignal,'single');
            fclose(fp);
            
            if(s.verbose)
                display('MAST:Pitch extraction using the RAPT algorithm...')
            end
            
            [status, result] = systemCallMAST(s, [ s.pitch ' -o 1 -a 0 -s ' num2str(s.fs/1000) ' -p ' num2str(s.frameShift) ' -L ' num2str(s.f0Min) ' -H ' num2str(s.f0Max) ' ' rawFile ' > ' f0File] );
            if(status)
                error(result);
            end
            
            %Prepare object to be returned
            pitch = pitchData(s);
            pitch = pitch.read(f0File,'single');
            pitch.sourceFilename = tdd.sourceFilename;
            pitch.fs = s.fs;
            
            
            %Clean up
            if(isunix())
                [status, result] = systemCallMAST(s,[ 'rm -f ' f0File ' ' rawFile ]); %remove temporary files
                %TODO: For Windows
            end
            if(status)
                error(result)
            end
            
            if(s.verbose)
                display('MAST:Pitch extraction using the RAPT algorithm complete.')
            end
            
        end
        
        function pitch = extractPitchSWIPE(s,tdd)
            %speechSignal values are floats in [0,1]

            
            if(~isa(tdd,'timeDomainData'))
                error('Input object is not of type timeDomainData')
            end

            f0File = [s.tmpDir filesep 'sptkExtractPitchSWIPE_f0'];
            rawFile = [s.tmpDir filesep 'sptkExtractPitchSWIPE_tmprawfile.raw'];
            
            speechSignal = tdd.data;
            if(size(speechSignal,2)==2)
                warning('Using column 1 of 2-column data') %#ok<WNTAG>
                speechSignal = speechSignal(:,1);
            end
            speechSignal = speechSignal ./ max(abs(speechSignal));
            
            fp = fopen(rawFile,'w');
            fwrite(fp,speechSignal,'double');
            fclose(fp);
            
            if(s.verbose)
                display('MAST:Pitch extraction using the SWIPE algorithm...')
            end
            
            [status, result] = systemCallMAST(s, [ s.pitch ' -a 1 -t ' num2str(s.thresholdSWIPE) ' -s ' num2str(s.fs/1000) ' -p ' num2str(s.frameShift) ' -L ' num2str(s.f0Min) ' -H ' num2str(s.f0Max) ' ' rawFile ' > ' f0File] );
            if(status)
                error(result);
            end
            
            %Prepare object to be returned
            pitch = pitchData(s);
            pitch = pitch.read(f0File,'double');
            pitch.sourceFilename = tdd.sourceFilename;
            pitch.fs = s.fs;
            
            
            %Clean up
            if(isunix())
                [status, result] = systemCallMAST(s,[ 'rm -f ' f0File ' ' rawFile ]); %remove temporary files
                %TODO: For Windows
            end
            if(status)
                error(result)
            end
            
            if(s.verbose)
                display('MAST:Pitch extraction using the SWIPE algorithm complete.')
            end
            
        end
        
       
        
        
        function pulse = exciteFromPitch(s,pd)
            %Usage 
            %pulse = s.exciteFromPitch(pitch)
            
            tmpPulsePeriodFile = [s.tmpDir filesep 'MAST_exciteFromPitch_tmppulseperiod.pp'];
            tmpPulseFile = [s.tmpDir filesep 'MAST_exciteFromPitch_tmppulse.pulse'];
            
            f0 = pd.data;
            pitchPeriods = (1./f0*s.fs);
            pitchPeriods(pitchPeriods==Inf) = 0;
            fp = fopen(tmpPulsePeriodFile,'w');
            if(fp)
                fwrite(fp,pitchPeriods,s.precision);
            else
                error([tmpPulsePeriodFile ' could not be opened for writing'])
            end
            
            if(s.verbose)
                display('MAST:Preparing pulse excitation from pitch trajectory using excite...')
            end
            
            [status result] = systemCallMAST(s,[s.excite ' -p ' num2str(pd.frameShift) ' -n ' tmpPulsePeriodFile ' > ' tmpPulseFile ]);
            if(status)
                error(result)
            end
            
            pulse = timeDomainData(s,tmpPulseFile);
            
            %Clean up
            [status result] = systemCallMAST(s,['rm ' tmpPulsePeriodFile ' ' tmpPulseFile]);
            if(status)
                error(result)
            end
            if(s.verbose)
                display('MAST:Preparing pulse excitation from pitch trajectory using excite complete')
            end
            
        end
        
        function [status result] = systemCallMAST(s,command)
            [status result] = system(command);
            if(s.speakAloud)
                display(['MAST: ' command])
            end
        end
        
    end
end
